// @atlaskit/rovo-triggers/dist/es2019/common/utils/params/index.js
import { ROVO_PARAM_PREFIX, ROVO_VALID_PARAMS } from './constants';
export const firstCharUpper = str => str.charAt(0).toUpperCase() + str.slice(1);
export const firstCharLower = str => str.charAt(0).toLowerCase() + str.slice(1);
const isValidURL = url => {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
};

// For new style camelCase params
export const addPrefix = param => `${ROVO_PARAM_PREFIX}${firstCharUpper(param)}`;
export const removePrefix = param => {
  return firstCharLower(param.replace(ROVO_PARAM_PREFIX, ''));
};

// Creates RovoChatParams from URLSearchParams
// Optionally filter to specific params
const processParams = (input, options = {
  filter: []
}) => {
  const output = {};
  const safeSearchParamsInput = typeof input === 'string' ? isValidURL(input) ? new URLSearchParams(new URL(input).search) : new URLSearchParams(input) : input;
  const processedInput = new URLSearchParams(safeSearchParamsInput);
  const extracted = new URLSearchParams();
  safeSearchParamsInput.forEach((value, key) => {
    var _options$filter;
    // Only look at rovoChat params
    if (!key.startsWith(ROVO_PARAM_PREFIX)) {
      return;
    }
    const paramKey = removePrefix(key);
    if (options !== null && options !== void 0 && (_options$filter = options.filter) !== null && _options$filter !== void 0 && _options$filter.length) {
      if (options.filter.includes(paramKey)) {
        output[paramKey] = decodeURIComponent(value);
        extracted.append(key, value);
      }
    } else {
      output[paramKey] = decodeURIComponent(value);
    }

    // Remove rovoParam from processed input
    processedInput.delete(key);
  });
  const combinedQueryString = [processedInput, extracted].map(params => params.toString()).filter(part => part.length > 0).join('&');
  return {
    processed: processedInput,
    rovoParams: output,
    combinedQueryString
  };
};

// Get all rovoChat params from a URL or the current window location if undefined
export const getRovoParams = url => {
  try {
    const search = url ? new URL(url).search : window.location.search;
    const q = new URLSearchParams(search);
    return processParams(q).rovoParams;
  } catch {
    return {};
  }
};

// Update the address bar without reloading the page
export const updatePageRovoParams = params => {
  window.history.pushState({}, '', addRovoParamsToUrl(window.location.pathname, params));
};

// Add any valid rovoChat params to a URL
export const addRovoParamsToUrl = (url, params) => {
  const {
    processed,
    rovoParams
  } = processParams(url);
  const rovoParamsWithExisting = {
    ...rovoParams,
    ...params
  };
  const baseUrl = url.includes('?') ? url.split('?')[0] : url;
  const baseQuery = Array.from(processed).length ? `?${processed.toString()}&` : '?';
  const updatedRovoParamsString = encodeRovoParams(rovoParamsWithExisting);
  return `${baseUrl}${baseQuery}${updatedRovoParamsString}`;
};
export const encodeRovoParams = (params, encodeAsObject) => {
  if (encodeAsObject) {
    return Object.entries(params).reduce((acc, [key, value]) => {
      acc[addPrefix(key)] = encodeURIComponent(value);
      return acc;
    }, {});
  }
  return Object.entries(params).filter(([_, value]) => value !== undefined).map(([key, value]) => {
    return `${addPrefix(key)}=${encodeURIComponent(value)}`;
  }).join('&');
};
export const assertOnlySpecificFieldsDefined = (params, fields) => {
  return Object.entries(params).every(([key, value]) => {
    const field = fields.find(field => field === key);
    if (field) {
      return value !== undefined;
    } else {
      return value === undefined;
    }
  });
};
export const getListOfRovoParams = ({
  resourceRouterQuery = false
} = {}) => {
  /*
  For products using react-resource-router (e.g Atlas) to
  pass as the value of `query` on the route definition. It ensures that our parameters
  are not stripped from the URL when loading the page, or through redirect hops
  Any new parameters we want to maintain should be added here as well, which means
  only a version bump is needed in those products to support the latest param list
  The !=false simply means that the parameter is supported but is optional, and should
  not contain a value of 'false'
  */
  const suffix = resourceRouterQuery ? '!=false' : '';
  return ROVO_VALID_PARAMS.map(param => `${addPrefix(param)}${suffix}`);
};
