// @atlaskit/adf-schema/dist/cjs/schema/nodes/list-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listItemWithLocalId = exports.listItemWithFlexibleFirstChildStage0 = exports.listItem = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _utils = require("../../utils");
/**
 * @name list_item
 * @description this node allows task-list to be nested inside list-item
 */
var listItem = exports.listItem = (0, _nodeTypes.listItem)({
  parseDOM: [{
    tag: 'li'
  }],
  toDOM: function toDOM() {
    return ['li', 0];
  }
});

/**
 * @name list_item_with_local_id
 * @description this node allows list items to have a localId attribute
 */
var listItemWithLocalId = exports.listItemWithLocalId = (0, _nodeTypes.listItem)({
  parseDOM: [{
    tag: 'li',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    return ['li', {
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) || undefined
    }, 0];
  }
});

/**
 * @name list_item_with_flexible_first_child_stage0
 * @description stage0 listItem with flexible first child (see EDITOR-5417)
 */
var listItemWithFlexibleFirstChildStage0 = exports.listItemWithFlexibleFirstChildStage0 = (0, _nodeTypes.listItemWithFlexibleFirstChildStage0)({
  parseDOM: [{
    tag: 'li',
    getAttrs: function getAttrs() {
      return {
        localId: _utils.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs2;
    return ['li', {
      'data-local-id': (node === null || node === void 0 ? void 0 : (_node$attrs2 = node.attrs) === null || _node$attrs2 === void 0 ? void 0 : _node$attrs2.localId) || undefined
    }, 0];
  }
});
