// @atlaskit/react-ufo/dist/es2019/config/index.js
import { fg } from '@atlaskit/platform-feature-flags';
let config;
export const DEFAULT_TTVC_REVISION = 'fy26.04';
export function getDefaultTTVCRevision() {
  return fg('ufo_update_and_enforce_ttvc_v4_default_version') ? DEFAULT_TTVC_REVISION : 'fy25.03';
}
export const UNKNOWN_INTERACTION_RATE = 1000;

// Defensively typed, since this is directly user-editable
// and they could delete empty members

export function setUFOConfig(newConfig) {
  var _newConfig$vc;
  // Handle edge cases with `enabledVCRevisions`
  const {
    enabledVCRevisions
  } = (_newConfig$vc = newConfig === null || newConfig === void 0 ? void 0 : newConfig.vc) !== null && _newConfig$vc !== void 0 ? _newConfig$vc : {};
  if (fg('ufo_update_and_enforce_ttvc_v4_default_version') ? enabledVCRevisions : typeof (enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.byExperience) === 'object') {
    var _enabledVCRevisions$a;
    const byExperience = typeof (enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.byExperience) === 'object' ? enabledVCRevisions.byExperience : {};
    config = {
      ...newConfig,
      vc: {
        ...newConfig.vc,
        enabledVCRevisions: {
          // enforce axiom about `enabledVCRevisions.all` config:
          // DEFAULT_TTVC_REVISION must always be present in `all`
          all: Array.from(new Set([getDefaultTTVCRevision(), ...((_enabledVCRevisions$a = enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.all) !== null && _enabledVCRevisions$a !== void 0 ? _enabledVCRevisions$a : []), ...Object.values(byExperience).flat()])),
          byExperience
        }
      }
    };
  } else {
    config = newConfig;
  }
}
export function getConfig() {
  return config;
}

/**
 * Check if UFO is enabled based on the config.enabled field.
 *
 * This function is gated behind the platform_ufo_enable_killswitch_config feature flag.
 * When the feature flag is disabled, UFO is always considered enabled (default behavior).
 * When the feature flag is enabled, the config.enabled field is respected.
 *
 * @returns true if UFO is enabled, false if disabled
 */
export function isUFOEnabled() {
  var _config;
  // Only respect the config.enabled field when the killswitch feature flag is enabled
  if (!fg('platform_ufo_enable_killswitch_config')) {
    return true;
  }
  // Default to enabled if config is not set or enabled is not explicitly false
  return ((_config = config) === null || _config === void 0 ? void 0 : _config.enabled) !== false;
}
const isValidConfigArray = array => {
  return Array.isArray(array) && array.length > 0;
};
export function getEnabledVCRevisions(experienceKey = '') {
  try {
    var _config$vc;
    if (!config) {
      return [];
    }
    if ((_config$vc = config.vc) !== null && _config$vc !== void 0 && _config$vc.enabled) {
      var _config$vc2, _enabledVCRevisions$b;
      const {
        enabledVCRevisions
      } = (_config$vc2 = config.vc) !== null && _config$vc2 !== void 0 ? _config$vc2 : {};
      if (isValidConfigArray(enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : (_enabledVCRevisions$b = enabledVCRevisions.byExperience) === null || _enabledVCRevisions$b === void 0 ? void 0 : _enabledVCRevisions$b[experienceKey])) {
        var _enabledVCRevisions$b2;
        return (_enabledVCRevisions$b2 = enabledVCRevisions.byExperience) === null || _enabledVCRevisions$b2 === void 0 ? void 0 : _enabledVCRevisions$b2[experienceKey];
      }
      if (isValidConfigArray(enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.all)) {
        return enabledVCRevisions.all;
      }
      return [getDefaultTTVCRevision()];
    }
    return [];
  } catch {
    return [];
  }
}
export function isVCRevisionEnabled(revision, experienceKey) {
  return getEnabledVCRevisions(experienceKey).includes(revision);
}
export function getMostRecentVCRevision(experienceKey = '') {
  const enabledVCRevisions = getEnabledVCRevisions(experienceKey);
  return enabledVCRevisions[enabledVCRevisions.length - 1];
}
export function getReactHydrationStats() {
  if (!config) {
    return undefined;
  }
  try {
    var _config2, _config2$getReactHydr;
    const stats = (_config2 = config) === null || _config2 === void 0 ? void 0 : (_config2$getReactHydr = _config2.getReactHydrationStats) === null || _config2$getReactHydr === void 0 ? void 0 : _config2$getReactHydr.call(_config2);
    return stats !== null && stats !== void 0 ? stats : undefined;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('getReactHydrationStats threw an error: ', e);
    return undefined;
  }
}
export function getInteractionRate(name, interactionKind) {
  try {
    if (!config) {
      return 0;
    }
    const {
      killswitch,
      rates,
      rules,
      kind,
      autoGeneratedRate
    } = config;
    if (killswitch != null) {
      // Specifically kill certain events
      if (killswitch.includes(name)) {
        return 0;
      }
    }
    const isUnknown = name === 'unknown';
    if (isUnknown) {
      if ((rates === null || rates === void 0 ? void 0 : rates[name]) != null) {
        return rates[name];
      }
      // Return default rate for unknown events when no explicit rate is configured
      return UNKNOWN_INTERACTION_RATE;
    }
    if (rates != null) {
      // Decide the rate explicitly
      if (rates[name] != null) {
        return rates[name];
      }
    }
    if (rules != null) {
      for (let i = 0; i < rules.length; i++) {
        const rule = rules[i];
        const {
          test,
          rate
        } = rule;
        if (test != null && rate != null) {
          const regexp = new RegExp(test, 'ig');
          if (regexp.test(name)) {
            return rate;
          }
        }
      }
    }
    const isAutogenerated = name === null || name === void 0 ? void 0 : name.startsWith('auto-generated');
    if (isAutogenerated) {
      const rate = autoGeneratedRate;
      if (rate != null) {
        return rate;
      }
    }
    if (kind != null) {
      const rate = kind[interactionKind];
      if (rate != null) {
        return rate;
      }
    }
    return 0;
  } catch {
    // Fallback
    return 0;
  }
}
export function getExperimentalInteractionRate(name, interactionType) {
  try {
    if (!config) {
      return 0;
    }
    const {
      experimentalInteractionMetrics
    } = config;
    if (!(experimentalInteractionMetrics !== null && experimentalInteractionMetrics !== void 0 && experimentalInteractionMetrics.enabled)) {
      return 0;
    }
    if (experimentalInteractionMetrics.rates && typeof experimentalInteractionMetrics.rates[name] === 'number') {
      return experimentalInteractionMetrics.rates[name];
    }
    if (experimentalInteractionMetrics.kind && typeof experimentalInteractionMetrics.kind[interactionType] === 'number') {
      return experimentalInteractionMetrics.kind[interactionType];
    }
    return 0;
  } catch {
    return 0;
  }
}
export function getVCRawDataInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'enableVCRawDataRates');
}
export function shouldUseRawDataThirdPartyBehavior(name, interactionType) {
  if (!fg('platform_ufo_raw_data_thirdparty')) {
    return false;
  }
  return getVCRawDataInteractionRate(name, interactionType) > 0;
}
export function getPostInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'postInteractionLog');
}
export function getCapabilityRate(capability) {
  if (fg('platform_ufo_remove_deprecated_config_fields')) {
    return 1;
  }
  try {
    if (!config) {
      return 0;
    }
    const {
      capability: capabilityRate
    } = config;
    if (capabilityRate != null) {
      const rate = capabilityRate[capability];
      if (rate != null) {
        return rate;
      }
    }
    return 0;
  } catch {
    return 0;
  }
}
function getConfigRate(name, interactionType, configName) {
  try {
    if (!config) {
      return 0;
    }
    const gotConfig = config[configName];
    if (!(gotConfig !== null && gotConfig !== void 0 && gotConfig.enabled)) {
      return 0;
    }
    if (interactionType !== 'page_load' && interactionType !== 'transition') {
      return 0;
    }
    if (gotConfig.rates && typeof gotConfig.rates[name] === 'number') {
      return gotConfig.rates[name];
    }
    if ('kind' in gotConfig && gotConfig.kind && typeof gotConfig.kind[interactionType] === 'number') {
      return gotConfig.kind[interactionType];
    }
    return 0;
  } catch {
    return 0;
  }
}
export function getExtraInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'extraInteractionMetrics');
}
const validTypingMethods = ['timeout', 'timeoutNoAlloc', 'mutationObserver'];
export function getTypingPerformanceTracingMethod() {
  const defaultMethod = 'timeout';
  try {
    if (!config) {
      return defaultMethod;
    }
    const {
      typingMethod
    } = config;
    if (typingMethod != null && validTypingMethods.find(m => m === typingMethod)) {
      return typingMethod;
    } else {
      return defaultMethod;
    }
  } catch {
    return defaultMethod;
  }
}

// Will ask UFO to wait for BM3 TTI for all events on this list
export function getAwaitBM3TTIList() {
  try {
    if (!config) {
      return [];
    }
    const {
      awaitBM3TTI
    } = config;
    if (awaitBM3TTI != null) {
      return awaitBM3TTI;
    } else {
      return [];
    }
  } catch {
    return [];
  }
}

// Will ask UFO to override ufoName => experienceKey to the apdex key
// This is to cover use cases where one rUFO route is split into multiple BM3 keys (1:N)
// This workaround should only be used as last resource,
// when we find router limitations preventing teams from splliting these routes
export function getUfoNameOverrides() {
  try {
    if (!config) {
      return undefined;
    }
    const {
      ufoNameOverrides
    } = config;
    if (ufoNameOverrides != null) {
      return ufoNameOverrides;
    }
    return undefined;
  } catch {
    return undefined;
  }
}
export function getMinorInteractions() {
  try {
    if (!config) {
      return undefined;
    }
    const {
      minorInteractions
    } = config;
    return minorInteractions;
  } catch {
    return undefined;
  }
}

// Contains the list of interactions that do not abort existing known interactions
export function getDoNotAbortActivePressInteraction() {
  try {
    if (!config) {
      return undefined;
    }
    const {
      doNotAbortActivePressInteraction
    } = config;
    return doNotAbortActivePressInteraction;
  } catch {
    return undefined;
  }
}

// Contains the list of interactions that do not abort on transitions
export function getDoNotAbortActivePressInteractionOnTransition() {
  try {
    if (!config) {
      return undefined;
    }
    const {
      doNotAbortActivePressInteractionOnTransition
    } = config;
    return doNotAbortActivePressInteractionOnTransition;
  } catch {
    return undefined;
  }
}

// Contains the list of interactions that do not abort on transitions
export function getFinishInteractionOnTransition() {
  try {
    if (!config) {
      return undefined;
    }
    const {
      finishInteractionOnTransition
    } = config;
    return finishInteractionOnTransition;
  } catch {
    return undefined;
  }
}
export const CLEANUP_TIMEOUT = 60 * 1000;
export function getInteractionTimeout(ufoName) {
  try {
    if (!config) {
      return CLEANUP_TIMEOUT;
    }
    const {
      interactionTimeout
    } = config;
    if (interactionTimeout != null && interactionTimeout[ufoName] != null) {
      return interactionTimeout[ufoName];
    }
    if (interactionTimeout != null && interactionTimeout.__globalInteractionTimeout != null) {
      return interactionTimeout.__globalInteractionTimeout;
    } else {
      return CLEANUP_TIMEOUT;
    }
  } catch {
    return CLEANUP_TIMEOUT;
  }
}
