// @atlaskit/react-ufo/dist/esm/vc/vc-observer-new/metric-calculator/utils/detect-layout-shift-cause.js
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
var calculateOffenderState = function calculateOffenderState(current, matchesCurrentEntry) {
  // Once we know it's mixed across entries, we can short-circuit.
  if (current === 'some') {
    return 'some';
  }
  var entryState = matchesCurrentEntry ? 'all' : 'none';
  if (current == null) {
    return entryState;
  }

  // If the state flips between entries (all <-> none), it's "some".
  if (current !== entryState) {
    return 'some';
  }
  return current;
};
var calculateLayoutShiftMovement = function calculateLayoutShiftMovement(layoutShiftEntries) {
  var movements = layoutShiftEntries.map(function (lsEntry) {
    var rect = lsEntry.rect;
    var previousRect = lsEntry.previousRect;
    if (!rect || !previousRect) {
      return null;
    }
    var dx = rect.x - previousRect.x;
    var dy = rect.y - previousRect.y;
    return {
      dx: dx,
      dy: dy,
      movedX: dx !== 0,
      movedY: dy !== 0,
      dirX: Math.sign(dx),
      dirY: Math.sign(dy),
      absDx: Math.abs(dx),
      absDy: Math.abs(dy)
    };
  });
  var hasMovement = function hasMovement(m) {
    return m !== null;
  };
  var validMovements = movements.filter(hasMovement);
  var allHaveRects = layoutShiftEntries.length > 0 && validMovements.length === layoutShiftEntries.length;
  if (!allHaveRects) {
    return {
      allHaveRects: allHaveRects,
      allMovedSameWay: false,
      allMovedSameAmount: false
    };
  }
  var first = validMovements[0];
  var allMovedSameWay = validMovements.every(function (m) {
    return m.movedX === first.movedX && m.movedY === first.movedY && m.dirX === first.dirX && m.dirY === first.dirY;
  });
  var allMovedSameAmount = allMovedSameWay && validMovements.every(function (m) {
    return m.absDx === first.absDx && m.absDy === first.absDy;
  });
  return {
    allHaveRects: allHaveRects,
    allMovedSameWay: allMovedSameWay,
    allMovedSameAmount: allMovedSameAmount,
    deltaX: first.dx,
    deltaY: first.dy
  };
};
export var detectLayoutShiftCause = function detectLayoutShiftCause(_ref) {
  var viewportEntries = _ref.viewportEntries,
    layoutShiftEntries = _ref.layoutShiftEntries,
    time = _ref.time,
    startTime = _ref.startTime,
    _ref$offenderWindowMs = _ref.offenderWindowMs,
    offenderWindowMs = _ref$offenderWindowMs === void 0 ? 250 : _ref$offenderWindowMs;
  var checkpointTimestamp = startTime + time;
  var filteredLSPotentialOffenders = viewportEntries.filter(function (viewportEntry) {
    return (viewportEntry.time < checkpointTimestamp && viewportEntry.time > checkpointTimestamp - offenderWindowMs || viewportEntry.time > checkpointTimestamp && viewportEntry.time < checkpointTimestamp + offenderWindowMs) && viewportEntry.data.type !== 'layout-shift';
  });

  // Summarize whether all layout-shift entries moved in a consistent direction and by the same amount.
  var layoutShiftVariables = calculateLayoutShiftMovement(layoutShiftEntries);

  // Classify each offender against *all* layout-shift entries when the layout shift is consistently moving in a
  // single axis (pure X or pure Y) so we can reason about whether an offender could have caused the movement.
  var layoutShiftOffenders = filteredLSPotentialOffenders.reduce(function (acc, offender) {
    var _offenderData$origina;
    var offenderData = offender.data;
    var offenderRect = offenderData.rect;
    if (!offenderRect) {
      return acc;
    }
    var offenderTimestamp = (_offenderData$origina = offenderData.originalMutationTimestamp) !== null && _offenderData$origina !== void 0 ? _offenderData$origina : offender.time;
    var happenedBefore = offenderTimestamp <= checkpointTimestamp;
    var offenderLeft = offenderRect.x;
    var offenderRight = offenderRect.x + offenderRect.width;
    var offenderTop = offenderRect.y;
    var offenderBottom = offenderRect.y + offenderRect.height;
    var shouldClassifyAgainstAllEntries = layoutShiftVariables.allMovedSameWay && 'deltaX' in layoutShiftVariables && 'deltaY' in layoutShiftVariables && layoutShiftVariables.deltaX === 0 !== (layoutShiftVariables.deltaY === 0);
    var isAbove = null;
    var isLeft = null;
    var isRight = null;
    var hasHorizontalOverlap = null;
    var hasVerticalOverlap = null;
    if (shouldClassifyAgainstAllEntries) {
      var _iterator = _createForOfIteratorHelper(layoutShiftEntries),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var lsEntry = _step.value;
          var lsRect = lsEntry.rect;
          if (!lsRect) {
            continue;
          }
          var lsLeft = lsRect.x;
          var lsRight = lsRect.x + lsRect.width;
          var lsTop = lsRect.y;
          var lsBottom = lsRect.y + lsRect.height;
          isAbove = calculateOffenderState(isAbove, offenderBottom <= lsTop);
          isLeft = calculateOffenderState(isLeft, offenderRight <= lsLeft);
          isRight = calculateOffenderState(isRight, offenderLeft >= lsRight);
          hasHorizontalOverlap = calculateOffenderState(hasHorizontalOverlap, offenderLeft < lsRight && offenderRight > lsLeft);
          hasVerticalOverlap = calculateOffenderState(hasVerticalOverlap, offenderTop < lsBottom && offenderBottom > lsTop);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    var matchesLayoutShiftDelta = false;
    if (shouldClassifyAgainstAllEntries && layoutShiftVariables.allMovedSameAmount && offenderData.previousRect) {
      var dx = offenderRect.x - offenderData.previousRect.x;
      var dy = offenderRect.y - offenderData.previousRect.y;
      var isPureX = layoutShiftVariables.deltaX !== 0 && layoutShiftVariables.deltaY === 0;
      var isPureY = layoutShiftVariables.deltaY !== 0 && layoutShiftVariables.deltaX === 0;
      if (isPureX) {
        matchesLayoutShiftDelta = dx === layoutShiftVariables.deltaX && dy === 0;
      } else if (isPureY) {
        matchesLayoutShiftDelta = dy === layoutShiftVariables.deltaY && dx === 0;
      }
    }
    acc.push({
      offender: offenderData.elementName,
      happenedBefore: happenedBefore,
      distanceToLS: Math.round(offenderTimestamp - checkpointTimestamp),
      isAbove: isAbove !== null && isAbove !== void 0 ? isAbove : 'none',
      isLeft: isLeft !== null && isLeft !== void 0 ? isLeft : 'none',
      isRight: isRight !== null && isRight !== void 0 ? isRight : 'none',
      hasHorizontalOverlap: hasHorizontalOverlap !== null && hasHorizontalOverlap !== void 0 ? hasHorizontalOverlap : 'none',
      hasVerticalOverlap: hasVerticalOverlap !== null && hasVerticalOverlap !== void 0 ? hasVerticalOverlap : 'none',
      matchesLayoutShiftDelta: matchesLayoutShiftDelta
    });
    return acc;
  }, []);
  return {
    layoutShiftVariables: layoutShiftVariables,
    layoutShiftOffenders: layoutShiftOffenders
  };
};
