// @atlaskit/react-ufo/dist/es2019/interaction-metrics-init/index.js
import { context } from '@opentelemetry/api';
import { fg } from '@atlaskit/platform-feature-flags';
import { startLighthouseObserver } from '../additional-payload';
import { isUFOEnabled, setUFOConfig } from '../config';
import { experimentalVC, sinkExperimentalHandler } from '../create-experimental-interaction-metrics-payload';
import { sinkExtraSearchPageInteractionHandler } from '../create-extra-search-page-interaction-payload';
import { setContextManager, UFOContextManager } from '../experience-trace-id-context/context-manager';
import { setupHiddenTimingCapture, setupThrottleDetection } from '../hidden-timing';
import { interactionExtraMetrics, postInteractionLog, sinkInteractionHandler, sinkPostInteractionLogHandler } from '../interaction-metrics';
import { getPerformanceObserver } from '../interactions-performance-observer';
import { initialiseMemoryObserver, initialisePressureObserver } from '../machine-utilisation';
import { sinkTerminalErrorHandler } from '../set-terminal-error';
import scheduleIdleCallback from './schedule-idle-callback';
let initialized = false;
function sinkInteraction(instance, payloadPackage) {
  function sinkFn(interactionId, interaction) {
    function onIdle() {
      payloadPackage.createPayloads(interactionId, interaction).then(payloads => {
        // NOTE: This API is used by the UFO DevTool Chrome Extension and Criterion
        const devToolObserver = globalThis.__ufo_devtool_onUfoPayload;
        payloads === null || payloads === void 0 ? void 0 : payloads.forEach(payload => {
          if (typeof devToolObserver === 'function') {
            devToolObserver === null || devToolObserver === void 0 ? void 0 : devToolObserver(payload);
          }
          instance.sendOperationalEvent(payload);
        });
      }).catch(error => {
        throw error;
      });
    }
    scheduleIdleCallback(onIdle);
  }
  sinkInteractionHandler(sinkFn);
}
function sinkExperimentalInteractionMetrics(instance, payloadPackage) {
  function experimentalMetricsSinkFn(interactionId, interaction) {
    function experimentalMetricsOnIdle() {
      const payloadPromise = payloadPackage.createExperimentalMetricsPayload(interactionId, interaction);
      payloadPromise.then(payload => {
        if (payload) {
          instance.sendOperationalEvent(payload);
        }
      });
    }
    scheduleIdleCallback(experimentalMetricsOnIdle);
  }
  sinkExperimentalHandler(experimentalMetricsSinkFn);
}
function sinkPostInteractionLog(instance, createPostInteractionLogPayload) {
  sinkPostInteractionLogHandler(logOutput => {
    scheduleIdleCallback(() => {
      const payload = createPostInteractionLogPayload(logOutput);
      if (payload) {
        instance.sendOperationalEvent(payload);
      }
    });
  });
}
function sinkTerminalErrors(instance, createTerminalErrorPayload) {
  sinkTerminalErrorHandler((errorData, context) => {
    scheduleIdleCallback(() => {
      const payload = createTerminalErrorPayload(errorData, context);
      if (payload) {
        instance.sendOperationalEvent(payload);
      }
    });
  });
}
function sinkInteractionExtraMetrics(instance, createInteractionExtraLogPayload) {
  interactionExtraMetrics.sinkHandler((interactionId, interaction, lastInteractionFinish, lastInteractionFinishVCResult) => {
    scheduleIdleCallback(async () => {
      const payload = await createInteractionExtraLogPayload(interactionId, interaction, lastInteractionFinish, lastInteractionFinishVCResult);
      if (payload) {
        instance.sendOperationalEvent(payload);
      }
    });
  });
}
function sinkExtraSearchPageInteraction(instance, payloadPackage) {
  function sinkFn(interactionId, interaction) {
    function onIdle() {
      payloadPackage.createExtraSearchPageInteractionPayload(interactionId, interaction).then(payloads => {
        // NOTE: This API is used by the UFO DevTool Chrome Extension and Criterion
        const devToolObserver = globalThis.__ufo_devtool_onUfoPayload;
        payloads === null || payloads === void 0 ? void 0 : payloads.forEach(payload => {
          if (typeof devToolObserver === 'function') {
            devToolObserver === null || devToolObserver === void 0 ? void 0 : devToolObserver(payload);
          }
          instance.sendOperationalEvent(payload);
        });
      }).catch(error => {
        throw error;
      });
    }
    scheduleIdleCallback(onIdle);
  }
  sinkExtraSearchPageInteractionHandler(sinkFn);
}
export function init(analyticsWebClientAsync, config) {
  var _config$vc;
  if (initialized) {
    return;
  }

  // Always set config first so isUFOEnabled() can check config.enabled
  setUFOConfig(config);

  // If UFO is disabled via config, skip all initialization
  // This is gated behind platform_ufo_enable_killswitch_config feature flag
  if (!isUFOEnabled()) {
    initialized = true;
    return;
  }

  // Initialize pressure observer for CPU usage monitoring
  initialisePressureObserver();
  initialiseMemoryObserver();
  if (fg('platform_ufo_enable_otel_context_manager')) {
    // Configure global OTel context manager
    const contextManager = new UFOContextManager();
    // set the contextmanager somewhere we can reference it later
    setContextManager(contextManager);
    // Register the context manager with the global OTel API
    contextManager.enable();
    context.setGlobalContextManager(contextManager);
  }
  if ((_config$vc = config.vc) !== null && _config$vc !== void 0 && _config$vc.enabled) {
    var _config$experimentalI, _config$extraInteract;
    const vcOptions = {
      heatmapSize: config.vc.heatmapSize,
      oldDomUpdates: config.vc.oldDomUpdates,
      devToolsEnabled: config.vc.devToolsEnabled,
      selectorConfig: config.vc.selectorConfig,
      ssrEnablePageLayoutPlaceholder: config.vc.ssrEnablePageLayoutPlaceholder,
      trackLayoutShiftOffenders: config.vc.trackLayoutShiftOffenders
    };
    postInteractionLog.initializeVCObserver(vcOptions);
    if (config !== null && config !== void 0 && (_config$experimentalI = config.experimentalInteractionMetrics) !== null && _config$experimentalI !== void 0 && _config$experimentalI.enabled) {
      experimentalVC.initialize(vcOptions).start({
        startTime: 0
      });
    }
    if (config !== null && config !== void 0 && (_config$extraInteract = config.extraInteractionMetrics) !== null && _config$extraInteract !== void 0 && _config$extraInteract.enabled) {
      interactionExtraMetrics.initializeVCObserver(vcOptions);
    }
  }
  setupHiddenTimingCapture();
  startLighthouseObserver();
  setupThrottleDetection();
  initialized = true;
  if (typeof PerformanceObserver !== 'undefined') {
    const observer = getPerformanceObserver();
    observer.observe({
      type: 'event',
      buffered: true,
      durationThreshold: 16
    });
  }
  Promise.all([analyticsWebClientAsync, import( /* webpackChunkName: "create-payloads" */'../create-payload'), import( /* webpackChunkName: "create-post-interaction-log-payload" */'../create-post-interaction-log-payload'), import( /* webpackChunkName: "create-interaction-extra-metrics-payload" */'../create-interaction-extra-metrics-payload'), import( /* webpackChunkName: "create-terminal-error-payload@atlaskit-internal_terminal_errors" */'../create-terminal-error-payload')]).then(([awc, payloadPackage, createPostInteractionLogPayloadPackage, createInteractionExtraMetricsPayloadPackage, createTerminalErrorPayloadPackage]) => {
    if (awc.getAnalyticsWebClientPromise) {
      awc.getAnalyticsWebClientPromise().then(client => {
        var _config$terminalError, _config$experimentalI2, _config$postInteracti, _config$extraInteract2, _config$extraSearchPa;
        const instance = client.getInstance();
        sinkInteraction(instance, payloadPackage);
        if (config !== null && config !== void 0 && (_config$terminalError = config.terminalErrors) !== null && _config$terminalError !== void 0 && _config$terminalError.enabled && fg('platform_ufo_enable_terminal_errors')) {
          sinkTerminalErrors(instance, createTerminalErrorPayloadPackage.default);
        }
        if (config !== null && config !== void 0 && (_config$experimentalI2 = config.experimentalInteractionMetrics) !== null && _config$experimentalI2 !== void 0 && _config$experimentalI2.enabled) {
          sinkExperimentalInteractionMetrics(instance, payloadPackage);
        }
        if ((_config$postInteracti = config.postInteractionLog) !== null && _config$postInteracti !== void 0 && _config$postInteracti.enabled) {
          sinkPostInteractionLog(instance, createPostInteractionLogPayloadPackage.default);
        }
        if (config !== null && config !== void 0 && (_config$extraInteract2 = config.extraInteractionMetrics) !== null && _config$extraInteract2 !== void 0 && _config$extraInteract2.enabled && fg('platform_ufo_enable_ttai_with_3p')) {
          sinkInteractionExtraMetrics(instance, createInteractionExtraMetricsPayloadPackage.default);
        }
        if (config !== null && config !== void 0 && (_config$extraSearchPa = config.extraSearchPageInteraction) !== null && _config$extraSearchPa !== void 0 && _config$extraSearchPa.enabled) {
          sinkExtraSearchPageInteraction(instance, payloadPackage);
        }
      });
    } else if (awc.sendOperationalEvent) {
      var _config$terminalError2, _config$experimentalI3, _config$postInteracti2, _config$extraInteract3, _config$extraSearchPa2;
      sinkInteraction(awc, payloadPackage);
      if (config !== null && config !== void 0 && (_config$terminalError2 = config.terminalErrors) !== null && _config$terminalError2 !== void 0 && _config$terminalError2.enabled && fg('platform_ufo_enable_terminal_errors')) {
        sinkTerminalErrors(awc, createTerminalErrorPayloadPackage.default);
      }
      if (config !== null && config !== void 0 && (_config$experimentalI3 = config.experimentalInteractionMetrics) !== null && _config$experimentalI3 !== void 0 && _config$experimentalI3.enabled) {
        sinkExperimentalInteractionMetrics(awc, payloadPackage);
      }
      if ((_config$postInteracti2 = config.postInteractionLog) !== null && _config$postInteracti2 !== void 0 && _config$postInteracti2.enabled) {
        sinkPostInteractionLog(awc, createPostInteractionLogPayloadPackage.default);
      }
      if (config !== null && config !== void 0 && (_config$extraInteract3 = config.extraInteractionMetrics) !== null && _config$extraInteract3 !== void 0 && _config$extraInteract3.enabled && fg('platform_ufo_enable_ttai_with_3p')) {
        sinkInteractionExtraMetrics(awc, createInteractionExtraMetricsPayloadPackage.default);
      }
      if (config !== null && config !== void 0 && (_config$extraSearchPa2 = config.extraSearchPageInteraction) !== null && _config$extraSearchPa2 !== void 0 && _config$extraSearchPa2.enabled) {
        sinkExtraSearchPageInteraction(awc, payloadPackage);
      }
    }
  });
}
